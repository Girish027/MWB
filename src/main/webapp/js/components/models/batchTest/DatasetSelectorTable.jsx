import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Table, Checkbox } from '@tfs/ui-components';
import { getFormattedDate } from 'utils/DateUtils';
import store from 'state/configureStore';
import * as actionsTagDatasets from 'state/actions/actions_tag_datasets';
import { RouteNames } from 'utils/routeHelpers';
import { changeRoute } from 'state/actions/actions_app';

class DatasetSelectorTable extends Component {
  constructor(props, context) {
    super(props, context);

    this.state = {
      tableData: [],
      availableDatasets: [],
      selectedDatasets: [],
    };

    this.onToggleDatasetCheckbox = this.onToggleDatasetCheckbox.bind(this);
    this.toggleSelectAll = this.toggleSelectAll.bind(this);
    this.updateSelectedDatasets = this.updateSelectedDatasets.bind(this);
  }

  static getDerivedStateFromProps(props, state) {
    let tableData = props.datasets ? Array.from(props.datasets.values()) : [];
    // TODO: revisit validation logic for datasets. Move it to Utils to be used commonly.
    tableData = tableData.filter(item => item.id && item.status && item.status == 'COMPLETED' && item.name && item.isClickable);
    if (state.availableDatasets.length === 0) {
      const availableDatasets = tableData.map(dataset => dataset.id);
      return ({
        availableDatasets,
        tableData,
      });
    }
    return ({
      tableData,
    });
  }

  updateSelectedDatasets(selectedDatasets) {
    this.setState({
      selectedDatasets,
    }, () => {
      this.props.updateSelectedDatasets(selectedDatasets);
    });
  }

  onToggleDatasetCheckbox(datasetId, checked) {
    let { selectedDatasets } = this.state;
    if (checked == true) {
      selectedDatasets = [...selectedDatasets, datasetId];
    } else {
      selectedDatasets = selectedDatasets.filter(id => id !== datasetId);
    }
    this.updateSelectedDatasets(selectedDatasets);
  }

  toggleSelectAll(checked) {
    let selectedDatasets = [];
    if (checked) {
      selectedDatasets = this.state.availableDatasets;
    }
    this.updateSelectedDatasets(selectedDatasets);
  }

  render() {
    // data available = {_key, id, client, projectId, name, type, description, locale, createdAt, status, task}
    const { selectedDatasets, availableDatasets } = this.state;
    const columns = [
      {
        Header: row => (
          <Checkbox
            onChange={(event) => {
              this.toggleSelectAll(event.target.checked);
            }}
            checked={(selectedDatasets.length > 0 && selectedDatasets.length === availableDatasets.length)}
          />
        ),
        id: 'checkbox',
        accessor: '',
        Cell: row => (
          <Checkbox
            onChange={(event) => {
              this.onToggleDatasetCheckbox(row.value.id, event.target.checked);
            }}
            checked={this.state.selectedDatasets.includes(row.value.id)}
          />
        ),
        maxWidth: 100,
        sortable: false,
      }, {
        Header: 'Dataset Names',
        accessor: '',
        id: 'name',
        Cell: ({ original }) => (original.isClickable ? (
          <a
            href="javascript:;"
            onClick={() => {
              const { history, client, projectId } = this.props;
              store.dispatch(actionsTagDatasets.setIncomingFilter({ projectId, datasets: [original.id] }));
              store.dispatch(changeRoute(RouteNames.TAG_DATASETS, { client, projectId }, history));
            }}
          >
            {original.name}
          </a>
        ) : (
          <div>
            {' '}
            {original.name}
            {' '}
          </div>
        )),

      }, {
        Header: 'User Id',
        accessor: row => row.modifiedBy,
        id: 'userId',
      }, {
        Header: 'Date Created',
        accessor: row => getFormattedDate(row.createdAt),
        id: 'createdAt',
      },
    ];


    return (
      <div>
        <Table
          data={this.state.tableData}
          columns={columns}
          defaultSorted={[{ id: 'createdAt', desc: true }]}
          showPagination={false}
          defaultPageSize={1000}
        />
      </div>
    );
  }
}

DatasetSelectorTable.propTypes = {
  history: PropTypes.object.isRequired,
  client: PropTypes.object.isRequired,
  projectId: PropTypes.string.isRequired,
  datasets: PropTypes.object,
  updateSelectedDatasets: PropTypes.func,
};

DatasetSelectorTable.defaultProps = {
  datasets: {},
  updateSelectedDatasets: () => {},
};

export default DatasetSelectorTable;
