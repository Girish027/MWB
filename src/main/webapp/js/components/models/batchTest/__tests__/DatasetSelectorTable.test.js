import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import DatasetSelectorTable from 'components/models/batchTest/DatasetSelectorTable';

describe('<DatasetSelectorTable />', () => {
  const createDataset = (id, name, createdAt, status = 'COMPLETED') => ({
    _key: id,
    id,
    clientId: 1,
    projectId: 2,
    name,
    type: 'Audio/Voice (Data Collection)',
    description: 'test dataset',
    locale: 'en_US',
    createdAt,
    status,
    task: 'INDEX',
    isClickable: true,
  });

  let datasets;
  let props;
  let wrapper;
  let expectedData;

  beforeAll(() => {
    datasets = new Map();
    datasets.set(1, createDataset(1, 'dataset 1', 1536777405845));
    datasets.set(2, createDataset(2, 'dataset 2', 1536777805845));
    datasets.set(3, createDataset(3, 'dataset 3', 1536778005208));

    expectedData = [
      createDataset(1, 'dataset 1', 1536777405845),
      createDataset(2, 'dataset 2', 1536777805845),
      createDataset(3, 'dataset 3', 1536778005208),
    ];

    props = {
      history: {},
      client: {},
      projectId: '123',
      datasets,
      updateSelectedDatasets: () => {},
    };
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  describe('Creating an instance', () => {
    test('should exist', () => {
      wrapper = shallow(<DatasetSelectorTable
        {...props}
      />);
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('Snapshots', () => {
    beforeEach(() => {
      wrapper = shallow(<DatasetSelectorTable
        {...props}
      />);
    });

    test('renders correctly with given datasets', () => {
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe('Functionality:', () => {
    let commonData = {};

    beforeAll(() => {
      props.updateSelectedDatasets = jest.fn();
    });

    beforeEach(() => {
      wrapper = shallow(<DatasetSelectorTable
        {...props}
      />);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    describe('getDerivedStateFromProps:', () => {
      test('should populate the list of available datasets', () => {
        const state = DatasetSelectorTable.getDerivedStateFromProps(props, { availableDatasets: [] });
        expect(state.availableDatasets).toEqual([1, 2, 3]);
      });

      test('should populate the list of available datasets which has COMPLETED as status', () => {
        const datasetsWithDiffStatus = new Map();
        datasetsWithDiffStatus.set(1, createDataset(1, 'dataset 1', 1536777405845, 'RUNNING'));
        datasetsWithDiffStatus.set(2, createDataset(2, 'dataset 2', 1536777805845, 'FAILED'));
        datasetsWithDiffStatus.set(3, createDataset(3, 'dataset 3', 1536778005208));

        const testProps = {
          datasets: datasetsWithDiffStatus,
          updateSelectedDatasets: () => {},
        };
        const state = DatasetSelectorTable.getDerivedStateFromProps(testProps, { availableDatasets: [] });
        expect(state.availableDatasets).toEqual([3]);
      });

      test('should populate the tableData', () => {
        const state = DatasetSelectorTable.getDerivedStateFromProps(props, { availableDatasets: [] });
        expect(state.tableData).toEqual(expectedData);
      });
    });

    describe('updateSelectedDatasets:', () => {
      test('should update the selected Datasets', () => {
        wrapper.instance().updateSelectedDatasets([1]);
        expect(wrapper.state().selectedDatasets).toEqual([1]);
      });

      test('should call the props updateSelectedDatasets to update the state of parent component', () => {
        wrapper.instance().updateSelectedDatasets([1]);
        expect(props.updateSelectedDatasets).toHaveBeenCalledWith([1]);
      });
    });

    describe('onToggleDatasetCheckbox:', () => {
      test('should add to the selected Datasets when the dataset is checked', () => {
        wrapper.setState({
          selectedDatasets: [],
        });
        wrapper.instance().onToggleDatasetCheckbox(1, true);
        expect(wrapper.state().selectedDatasets).toContain(1);
      });

      test('should remove from the selected Datasets when the dataset is unchecked', () => {
        wrapper.setState({
          selectedDatasets: [1, 2],
        });
        wrapper.instance().onToggleDatasetCheckbox(2, false);
        expect(wrapper.state().selectedDatasets).not.toContain(2);
      });
    });

    describe('toggleSelectAll:', () => {
      test('should add all the datasets when select all is checked', () => {
        wrapper.setState({
          selectedDatasets: [],
        });
        wrapper.instance().toggleSelectAll(true);
        expect(wrapper.state().selectedDatasets).toEqual([1, 2, 3]);
      });

      test('should remove all the datasets when selectAll is unchecked', () => {
        wrapper.setState({
          selectedDatasets: [1, 2],
        });
        wrapper.instance().toggleSelectAll(false);
        expect(wrapper.state().selectedDatasets).toEqual([]);
      });
    });
  });
});
