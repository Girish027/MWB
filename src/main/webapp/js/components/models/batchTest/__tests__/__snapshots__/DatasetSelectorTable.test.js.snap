// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatasetSelectorTable /> Snapshots renders correctly with given datasets 1`] = `
<div>
  <withTheme(Table)
    columns={
      Array [
        Object {
          "Cell": [Function],
          "Header": [Function],
          "accessor": "",
          "id": "checkbox",
          "maxWidth": 100,
          "sortable": false,
        },
        Object {
          "Cell": [Function],
          "Header": "Dataset Names",
          "accessor": "",
          "id": "name",
        },
        Object {
          "Header": "User Id",
          "accessor": [Function],
          "id": "userId",
        },
        Object {
          "Header": "Date Created",
          "accessor": [Function],
          "id": "createdAt",
        },
      ]
    }
    data={
      Array [
        Object {
          "_key": 1,
          "clientId": 1,
          "createdAt": 1536777405845,
          "description": "test dataset",
          "id": 1,
          "isClickable": true,
          "locale": "en_US",
          "name": "dataset 1",
          "projectId": 2,
          "status": "COMPLETED",
          "task": "INDEX",
          "type": "Audio/Voice (Data Collection)",
        },
        Object {
          "_key": 2,
          "clientId": 1,
          "createdAt": 1536777805845,
          "description": "test dataset",
          "id": 2,
          "isClickable": true,
          "locale": "en_US",
          "name": "dataset 2",
          "projectId": 2,
          "status": "COMPLETED",
          "task": "INDEX",
          "type": "Audio/Voice (Data Collection)",
        },
        Object {
          "_key": 3,
          "clientId": 1,
          "createdAt": 1536778005208,
          "description": "test dataset",
          "id": 3,
          "isClickable": true,
          "locale": "en_US",
          "name": "dataset 3",
          "projectId": 2,
          "status": "COMPLETED",
          "task": "INDEX",
          "type": "Audio/Voice (Data Collection)",
        },
      ]
    }
    defaultPageSize={1000}
    defaultSorted={
      Array [
        Object {
          "desc": true,
          "id": "createdAt",
        },
      ]
    }
    showPagination={false}
  />
</div>
`;
